### PR TITLE
[v1.8] NCL-5184 - Optimize queries to minimize CPU load and DB usage when hitting particular endpoints

### PR DESCRIPTION
**Pull request by vibe13**
…Records; collect all BuildConfigurationAudited entities in one query rather than looping

### Checklist:

* [ ] Have you added a note in the [CHANGELOG wiki](https://github.com/project-ncl/pnc/wiki/Changelog) for your change if user-facing?
* [ ] Have you added unit tests for your change?

@thescouser89 @matejonnet @jbartece Trying to optimize heavy endpoint

### CODE REVIEW

File: datastore/src/main/java/org/jboss/pnc/datastore/repositories/BuildConfigurationAuditedRepositoryImpl.java [hunk 1]
```diff
@@ -107,6 +111,43 @@ public BuildConfigurationAudited queryById(IdRev idRev) {
         );
     }
 
+    @Override
+    public Map<IdRev, BuildConfigurationAudited> queryById(Set<IdRev> idRevs) {
```
Comment (vibe13) [inline datastore/src/main/java/org/jboss/pnc/datastore/repositories/BuildConfigurationAuditedRepositoryImpl.java]: Query a set of BuildConfigurationAudited from a set of IdRevs

File: rest/src/main/java/org/jboss/pnc/rest/provider/BuildRecordProvider.java [hunk 2]
```diff
@@ -660,63 +696,76 @@ public BuildRecordRest getLatestBuildRecord(Integer configId, boolean executedOn
...
                 .map(p -> rsqlPredicateProducer.getPredicate(BuildRecord.class, p))
                 .collect(Collectors.toList());
         List<Predicate<BuildRecord>> dbOrPredicateList = new ArrayList<>();
 
+        // Combine rsqlQueries AND-ing all the single queries
         String combinedQueries = combineRsqlQueriesMatchAll(rsqlQueries);
-        if (!StringUtils.isEmpty(orFindByBuildConfigurationName)) {
-            //add steam condition
+
+        // An AND condition (andFindByBuildConfigurationName) gets higher precedence over an OR condition
```
Comment (vibe13) [inline rest/src/main/java/org/jboss/pnc/rest/provider/BuildRecordProvider.java]: If and AND condition (andFindByBuildConfigurationName) is specified, it gets processed. Else if an OR condition (orFindByBuildConfigurationName) is specified, it gets processed. If BOTH are specified (non sensible), make sure the only AND condition is picked.

File: rest/src/main/java/org/jboss/pnc/rest/provider/BuildRecordProvider.java [hunk 3]
```diff
@@ -660,63 +696,76 @@ public BuildRecordRest getLatestBuildRecord(Integer configId, boolean executedOn
...
-                    .searchForBuildConfigurationName(orFindByBuildConfigurationName);
-            if (!buildConfigurationAuditeds.isEmpty()) {
-                dbOrPredicateList.add(
-                        BuildRecordPredicates.withBuildConfigurationIdRev(
-                                buildConfigurationAuditeds.stream().map(bca -> bca.getIdRev()).collect(Collectors.toList())
-                        )
-                );
+            // add DB predicate
+            List<IdRev> buildConfigurationAuditedIdRevs = buildConfigurationAuditedRepository
+                    .searchIdRevForBuildConfigurationName(andFindByBuildConfigurationName);
```
Comment (vibe13) [inline rest/src/main/java/org/jboss/pnc/rest/provider/BuildRecordProvider.java]: The old code returned a list of BuildConfigurationAudited from their corresponding BuildConfiguration name, creating the BuildConfigurationAudited objects with an empty default BuildRecords list. It then used only the list of IdRev of the returned BuildConfigurationAudited list, to finally populate the BuildRecordPredicates list.

The new code returns directly the list of IdRev to be used to populate the BuildRecordPredicates list.

File: rest/src/main/java/org/jboss/pnc/rest/provider/BuildRecordProvider.java [hunk 4]
```diff
@@ -660,63 +696,76 @@ public BuildRecordRest getLatestBuildRecord(Integer configId, boolean executedOn
...
-                dbAndPredicatesList.add(
-                        BuildRecordPredicates.withBuildConfigurationIdRev(
-                                buildConfigurationAuditeds.stream().map(bca -> bca.getIdRev()).collect(Collectors.toList())
-                        )
-                );
-            } else {
-                dbAndPredicatesList.add(Predicate.nonMatching());
+            // add DB predicate
+            List<IdRev> buildConfigurationAuditedIdRevs = buildConfigurationAuditedRepository
+                    .searchIdRevForBuildConfigurationName(orFindByBuildConfigurationName);
```
Comment (vibe13) [inline rest/src/main/java/org/jboss/pnc/rest/provider/BuildRecordProvider.java]: Same as above comment.

File: rest/src/main/java/org/jboss/pnc/rest/provider/BuildRecordProvider.java [hunk 5]
```diff
@@ -660,63 +696,76 @@ public BuildRecordRest getLatestBuildRecord(Integer configId, boolean executedOn
...
             }
         }
 
-        Set<BuildRecordRest> running = nullableStreamOf(buildCoordinator.getSubmittedBuildTasks())
-                .map(this::createNewBuildRecordRest)
-                .filter(rsqlPredicateProducer.getStreamPredicate(BuildRecordRest.class, combinedQueries))
-                .sorted(sortInfoProducer.getSortInfo(sort).getComparator())
-                .collect(Collectors.toSet());
+        List<BuildTask> buildTasks = buildCoordinator.getSubmittedBuildTasks();
+        Set<IdRev> buildConfigAuditedIdRevsOfBuildTasks = nullableStreamOf(buildTasks).map(buildTask -> buildTask.getBuildConfigurationAudited().getIdRev()).collect(Collectors.toSet());
```
Comment (vibe13) [inline rest/src/main/java/org/jboss/pnc/rest/provider/BuildRecordProvider.java]: The old code invoked method createNewBuildRecordRest, which invoked **buildConfigurationAuditedRepository.queryById(IdRev)**. That method queried 1 by 1 all the BuildConfigurationAudited associated with eventually running BuildTasks; it queried the BuildRecords associated with each retrieved BuildConfiguration; it fetched the GenericParameters of each BuildConfiguration retrieved. Moreover, the BuildRecords queried, while used in inner methods to construct the BuildConfigurationAudited objects, were never used.
Considering **20** running builds, the number of queries done were:
20 x (1 +1 +1) = **60** <-- 20 running builds x (1 query for BuildConfigurationAudited; 1 query for all the BuildRecords; 1 query to fetch the GenericParameters)

The new code retrieves all the IdRevs of the BuildConfigurationAudited associated with the eventually running BuildTasks and retrieves with 1 query all the BuildConfigurationAudited objects; does not query the BuildRecords (not used anyway); queries the GenericParameters.
Thus, again considering **20** running builds, the number of queries done are: 
1 + 0 + 20 x 1 = **21** <-- 1 query for all 20 BuildConfigurationAudited; no query for BuildRecords; 1 query for each rBuildConfigurationAudited to fetch GenericParameters.

File: rest/src/main/java/org/jboss/pnc/rest/provider/BuildRecordProvider.java [hunk 6]
```diff
@@ -740,11 +789,18 @@ public BuildRecordRest getLatestBuildRecord(Integer configId, boolean executedOn
...
+        List<BuildRecord> buildRecords = nullableStreamOf(((BuildRecordRepository) repository)
                 .queryWithPredicatesUsingCursor(pageInfo, sortInfo, dbAndPredicates, dbOrPredicates))
-                .map(toRESTModel())
                 .collect(Collectors.toList());
+        Set<IdRev> buildRecordsIdRevs = nullableStreamOf(buildRecords).map(br -> new IdRev(br.getBuildConfigurationId(), br.getBuildConfigurationRev())).collect(Collectors.toSet());
+
+        List<BuildRecordRest> content = new ArrayList<>();
+        if (!buildRecordsIdRevs.isEmpty()) {
+            Map<IdRev, BuildConfigurationAudited> buildConfigurationsAuditedMap = buildConfigurationAuditedRepository.queryById(buildRecordsIdRevs);
+            content = nullableStreamOf(buildRecords).map(toRESTModel(buildConfigurationsAuditedMap))
```
Comment (vibe13) [inline rest/src/main/java/org/jboss/pnc/rest/provider/BuildRecordProvider.java]: Old method toRESTModel() invoked **buildConfigurationAuditedRepository.queryById(IdRev)**, so same comments as above are valid, and similar code changes applied.

**Comment by thescouser89** [comment]
Really sorry but I'm not too familiar with this part of the code. Will defer to others for the review. Sorry! :(

File: rest/src/main/java/org/jboss/pnc/rest/provider/BuildRecordProvider.java [hunk 7]
```diff
@@ -660,63 +696,76 @@ public BuildRecordRest getLatestBuildRecord(Integer configId, boolean executedOn
...
+        if (!StringUtils.isEmpty(andFindByBuildConfigurationName)) {
+            // add steam condition
+            StringBuilder sb = new StringBuilder(";(buildConfigurationName==").append(andFindByBuildConfigurationName).append(")");
+
             if (StringUtils.isEmpty(combinedQueries)) {
-                combinedQueries = "(buildConfigurationName=like=" + orFindByBuildConfigurationName + ")";
+                combinedQueries = sb.deleteCharAt(0).toString();
             } else {
-                combinedQueries = "(" + combinedQueries + "),(buildConfigurationName=like=" + orFindByBuildConfigurationName + ")";
+                combinedQueries = sb.insert(0, ")").insert(0, combinedQueries).insert(0, "(").toString();
```
Comment (janinko) [inline rest/src/main/java/org/jboss/pnc/rest/provider/BuildRecordProvider.java]: You don't need to use StringBuilder explicitly, the compiler will use it for you and the code will be nicer (e.g. not reversed order).

https://dzone.com/articles/string-concatenation-performacne-improvement-in-ja

File: rest/src/main/java/org/jboss/pnc/rest/provider/BuildRecordProvider.java [hunk 8]
```diff
@@ -660,63 +696,76 @@ public BuildRecordRest getLatestBuildRecord(Integer configId, boolean executedOn
...
+        } else if (!StringUtils.isEmpty(orFindByBuildConfigurationName)) {
+            // add steam condition
+            StringBuilder sb = new StringBuilder(",(buildConfigurationName=like=").append(orFindByBuildConfigurationName).append(")");
+
             if (StringUtils.isEmpty(combinedQueries)) {
-                combinedQueries = "(buildConfigurationName==" + andFindByBuildConfigurationName + ")";
+                combinedQueries = sb.deleteCharAt(0).toString();
             } else {
-                combinedQueries = "(" + combinedQueries + ");(buildConfigurationName==" + andFindByBuildConfigurationName + ")";
+                combinedQueries = sb.insert(0, ")").insert(0, combinedQueries).insert(0, "(").toString();
```
Comment (janinko) [inline rest/src/main/java/org/jboss/pnc/rest/provider/BuildRecordProvider.java]: same as above comment

File: datastore/src/main/java/org/jboss/pnc/datastore/repositories/BuildConfigurationAuditedRepositoryImpl.java [hunk 9]
```diff
@@ -107,6 +111,43 @@ public BuildConfigurationAudited queryById(IdRev idRev) {
...
+        logger.trace("Querying for BuildConfigurationAudited.idRevs: {}.", idRevs);
+
+        List<String> idRevConcatenated = idRevs.stream()
+                .map(idRev -> { return idRev.getId() + "-" + idRev.getRev();}).collect(Collectors.toList());
+
+        // WORKAROUND: as I cannot concatenate AuditEntity property to match `AuditEntity.property("id")-AuditEntity.property("rev")` in idRevConcatenated list
+        // I can query all BuildConfigurationAudited with the only id and later on filter id and rev
```
[the remark this answers is not in the document]
Comment (jbartece) [reply datastore/src/main/java/org/jboss/pnc/datastore/repositories/BuildConfigurationAuditedRepositoryImpl.java]: +1 AFAIK rev should be unique

[the remark this answers is not in the document]
Comment (vibe13) [reply datastore/src/main/java/org/jboss/pnc/datastore/repositories/BuildConfigurationAuditedRepositoryImpl.java]: Yes I have doubled checked, and it is indeed unique. Anyway, changed the query to seach for rev but also filtering on id (redundant but no uniqueness is enforced in db level by Hibernate)

File: datastore/src/main/java/org/jboss/pnc/datastore/repositories/BuildConfigurationAuditedRepositoryImpl.java [hunk 10]
```diff
@@ -107,6 +111,43 @@ public BuildConfigurationAudited queryById(IdRev idRev) {
         );
     }
 
+    @Override
+    public Map<IdRev, BuildConfigurationAudited> queryById(Set<IdRev> idRevs) {
+        logger.trace("Querying for BuildConfigurationAudited.idRevs: {}.", idRevs);
+
+        List<String> idRevConcatenated = idRevs.stream()
+                .map(idRev -> { return idRev.getId() + "-" + idRev.getRev();}).collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (vibe13) [reply datastore/src/main/java/org/jboss/pnc/datastore/repositories/BuildConfigurationAuditedRepositoryImpl.java]: Changed! Thanks!

File: rest/src/main/java/org/jboss/pnc/rest/provider/BuildRecordProvider.java (hunk 7 above)
Comment (vibe13) [reply rest/src/main/java/org/jboss/pnc/rest/provider/BuildRecordProvider.java]: Thanks for the hint, I guess I haven't been coding enough lately :-). And honestly my version looked so bad! :-)

**Pull request merged**